// @embroider/core@3.5.3/src/to-broccoli-plugin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = toBroccoliPlugin;
const broccoli_plugin_1 = __importDefault(require("broccoli-plugin"));
const shared_internals_1 = require("@embroider/shared-internals");
function toBroccoliPlugin(packagerClass) {
    class PackagerRunner extends broccoli_plugin_1.default {
        constructor(stage, variants, options) {
            super([stage.tree], {
                persistentOutput: true,
                needsCache: false,
                annotation: packagerClass.annotation,
            });
            this.stage = stage;
            this.variants = variants;
            this.options = options;
        }
        async build() {
            if (!this.packager) {
                this.packager = new packagerClass(this.stage.inputPath, this.outputPath, this.variants, msg => console.log(msg.split(shared_internals_1.tmpdir).join('$TMPDIR')), this.options);
            }
            return this.packager.build();
        }
    }
    return PackagerRunner;
}
